src/anchor_in_depth/Events.md

```diff
@@ -0,0 +1,92 @@
+
+Events in Anchor are basically base64 encoded structs that are program logged. For ex, when you may write msg!() something to debug your Solana program, you are logging data. Events are essentially the same thing but they are not human readable as they are base64 encoded. The reason for this is that it is quite expensive to log string data:
+
+```rust
+#[proc_macro]
+pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let data: proc_macro2::TokenStream = input.into();
+    proc_macro::TokenStream::from(quote! {
+        {
+            anchor_lang::solana_program::log::sol_log_data(&[&anchor_lang::Event::data(&#data)]);
+        }
+    })
+}
+```
+
+## Defining Events
+In Anchor Solana, events are defined using the `#[event]` attribute macro. This macro allows you to specify the fields that an event should contain. Events can include various data types, making them versatile for different use cases.
+
+```rust
+use anchor_lang::prelude::*;
+
+#[event]
+pub struct TransferEvent {
+    from: Pubkey,
+    to: Pubkey,
+    amount: u64,
+}
+```
+
+In this example, we define an event named `TransferEvent` with three fields: `from` (sender's address), `to` (receiver's address), and `amount` (the transferred amount).
+
+## Emitting Events
+To emit an event within your Anchor Solana program, you can use the `emit!` macro. This allows you to create an instance of the event structure and broadcast it to the network.
+
+```rust
+#[program]
+pub mod my_program {
+    use super::*;
+    
+    pub fn transfer(ctx: Context<TransferContext>, amount: u64) -> Result<()>  {
+        // Perform transfer logic
+        
+        // Emit the TransferEvent
+        emit!(TransferEvent {
+            from: *ctx.accounts.from.key,
+            to: *ctx.accounts.to.key,
+            amount,
+        });
+        
+        Ok(())
+    }
+}
+```
+
+In this example, when the `transfer` function is called, a `TransferEvent` is emitted using the `emit!` macro. The relevant data is populated into the event fields.
+
+## Subscribing to Events
+External parties can subscribe to events emitted by your smart contract using Solana's event subscription mechanisms. This enables them to listen for specific events and take actions accordingly.
+
+For example, using Anchor ts client library(@project-serum/anchor), you can subscribe to events like this:
+
+```javascript
+import * as anchor from '@project-serum/anchor'
+import { MyProgram, IDL } from 'my_program'
+const { web3 } = anchor
+let PROGRAM_ID = new web3.PublicKey("")
+const program = new anchor.Program(IDL, PROGRAM_ID) as Program<MyProgram>
+
+let listener = null;
+let [event, slot] = await new Promise((resolve, _reject) => {
+  listener = program.addEventListener("TransferEvent", (event, slot) => {
+    resolve([event, slot]);
+  });
+  program.rpc.initialize();
+});
+```
```

Comment:
There is a lot of stuff going on here that takes the focus out of the main important parts which is subscribing to events. It would also be better if we avoid the `Promise` stuff as that's just being used in tests and not realistically how events would be used.

```suggestion
```ts
const subscriptionId = program.addEventListener("TransferEvent", (event) => {
  // Handle event...
});
```
```